chain/chunks/src/lib.rs

```diff
@@ -1075,6 +1092,143 @@ impl ShardsManager {
+    ///
+    /// This requires encoding the chunk and as such is computationally
+    /// expensive operation.  If possible, the request should be served from
+    /// EncodedChunksCacheEntry or PartialEncodedChunk instead.
+    #[cfg_attr(feature = "test_features", visibility::make(pub))]
+    fn prepare_partial_encoded_chunk_response_from_chunk(
+        &mut self,
+        request: PartialEncodedChunkRequestMsg,
+        chain_store: &mut ChainStore,
+        rs: &mut ReedSolomonWrapper,
+        chunk: ShardChunk,
+    ) -> Option<PartialEncodedChunkResponseMsg> {
+        // TODO(mina86): Is it correct to use rs.data_shard_count() and
+        // rs.total_shard_count() here?  Can those values change over time?
+        let total_parts = rs.total_shard_count();
+        for ord in request.part_ords.iter() {
+            let ord: usize = (*ord).try_into().unwrap();
+            if ord >= total_parts {
+                debug!(target:"chunks", "Not sending {:?}, requested part_ord {} but we only expect {} total",
+                       request.chunk_hash.0, ord, total_parts);
+                return None;
+            }
+        }
+
+        // Figure out basic parameters such as shard id, epoch id, protocol
+        // version and shard layout.  Those will be needed later when fetching
+        // other data and constructing encoded shard chunk.
+        let header = chunk.cloned_header();
+        let prev_block_hash = header.prev_block_hash_ref();
+        let shard_id = header.shard_id();
+        let shard_layout = self
+            .runtime_adapter
+            .get_shard_layout_from_prev_block(prev_block_hash)
+            .map_err(|err| {
+                debug!(target: "chunks", "Not sending {:?}, failed to get shard layout: {}", request.chunk_hash.0, err);
+            })
+            .ok()?;
+
+        // Get outgoing receipts for the chunk and construct vector of their
+        // proofs.
+        let outgoing_receipts = chain_store
+            .get_outgoing_receipts_for_shard(
+                &*self.runtime_adapter,
+                *chunk.prev_block(),
+                shard_id,
+                chunk.height_included(),
+            )
+            .map_err(|err| {
+                debug!(target: "chunks", "Not sending {:?}, failed to get outgoing receipts: {}", request.chunk_hash.0, err);
+            }).ok()?;
+        let outgoing_receipts_hashes =
+            Chain::build_receipts_hashes(&outgoing_receipts, &shard_layout);
+        let (outgoing_receipts_root, outgoing_receipts_proofs) =
+            merklize(&outgoing_receipts_hashes);
+        if header.outgoing_receipts_root() != outgoing_receipts_root {
+            error!(target: "chunks",
+                   "Not sending {:?}, expected outgoing receipts root doesn’t match calculated: {} != {}",
+                   request.chunk_hash.0, header.outgoing_receipts_root(), outgoing_receipts_root);
+            return None;
+        }
+
+        let mut receipts_by_shard =
```

Comment:
This logic of constructing receipts_proofs has already appeared in both `distribute_encoded_chunk` and `create_and_persist_partial_chunk`, probably better to create a new method in `Chain` to encapsulate this logic to avoid code duplication